### PR TITLE
pkp/pkp-lib#8027 OPS preprints should support DOI versioning

### PR DESCRIPTION
**Pull request by ewhanson**

### CODE REVIEW

File: api/v1/dois/PKPDoiHandler.inc.php
Comment (NateWr) [inline api/v1/dois/PKPDoiHandler.inc.php:290]: Since `getPubObjectFromType()` and `updatePubObjectDoiFromType()` are all part of the funny logic of the API, let's make these protected methods of the handler class. I know you'll need a new `DoiHandler()` for each application, but it should be pretty small. You can also then define some constants for the types there which match the API specifications, without needing to roll this out to the rest of the application.

It might also free you to do a little bit more in the `edit` method and less in the helper methods. I would imagine something like a more general-purpose use function in the repository:

```php
class Repository
{
    public function isAssigned(int $doiId): bool
    {
        $publicationCount = ...;
        $galleyCount = ...;

        return $publicationCount || $galleyCount;
    }
}
```

And then the rest of the logic is in the API handler, and the app-specific API handler only needs to be responsible for returning the right variable for each object type. It doesn't have to take the action:

```php
class PKPDoiHandler extends APIHandler
{
    public const OBJECT_TYPE_SUBMISSION = 'submission';

    protected method getPubObjectRepo(string $type)
    {
        return match($type) {
            self::OBJECT_TYPE_SUBMISSION => Repo::submission(),
        };
    }

    public function edit(...)
    {
        // ...

        $repo = $this->getPubObjectRepo($type);
        if (!$repo) {
            return $response->withStatus(403)->withJsonError('pub.type.not.recognized');
        }
        
        // ...

        $repo->edit(...);

        if (!Repo::doi()->isAssigned($doi->getId()) {
            Repo::doi()->delete($doi);
        }
    }
}
```

File: classes/components/listPanels/PKPDoiListPanel.inc.php
Comment (NateWr) [inline classes/components/listPanels/PKPDoiListPanel.inc.php:172]: showing tabs not spaces here

Comment (NateWr) [reply classes/components/listPanels/PKPDoiListPanel.inc.php:172]: I'm noticing that throughout the changes in this PR...

File: locale/en_US/manager.po
Comment (NateWr) [inline locale/en_US/manager.po:851]: Can we get away with re-using existing locale keys, like "Save" and "Edit"?

File: classes/observers/listeners/VersionPublicationDois.php
Comment (NateWr) [inline classes/observers/listeners/VersionPublicationDois.php:60]: Reverse the logic here to return early:

```php
if (!$shouldVersionPublication || !$notFirstPublication) {
    return;
}
```

File: classes/observers/listeners/VersionPublicationDois.php
Comment (NateWr) [inline classes/observers/listeners/VersionPublicationDois.php:56]: The context is also in `PublishedEvent`:

```php
$context = $event->context;
```

File: classes/observers/listeners/VersionPublicationDois.php
Comment (NateWr) [inline classes/observers/listeners/VersionPublicationDois.php:59]: I'm not sure if this is certain to be true. I think it is possible to end up with a different version number even if it is the earliest publication that is currently published. It may be possible through publishing/unpublishing things.

You may need to rely on logic like you use below, where you check if a DOI exists that matches a DOI of a previous publication or something.

File: classes/observers/listeners/VersionPublicationDois.php
Comment (NateWr) [inline classes/observers/listeners/VersionPublicationDois.php:52]: I think you should maybe check if DOIs are enabled on the context-level before doing anything. When we have some settings (like should dois be versioned) which are dependent on other settings (dois enabled), there's always a risk that someone will have dois disabled but the old setting to version DOIs will still be in the system.

File: classes/observers/listeners/VersionPublicationDois.php
Comment (NateWr) [inline classes/observers/listeners/VersionPublicationDois.php:81]: Lines 62-81 would be good to move into a protected function. I'm not sure I understand all of the inline comments and one clear, succinct docblock above a function might make it easier to understand the general intent of what's happening.

File: classes/observers/listeners/VersionPublicationDois.php
Comment (NateWr) [inline classes/observers/listeners/VersionPublicationDois.php:45]: You use NB, i.e. and e.g. here. I'd prefer to just use words in these cases like "Please note" and "for example" or "such as".

File: classes/observers/listeners/VersionPublicationDois.php
Comment (NateWr) [reply classes/observers/listeners/VersionPublicationDois.php:81]: Similar with 83-101 below.

File: locale/en_US/manager.po
Comment (NateWr) [inline locale/en_US/manager.po:915]: > Most preprint servers should use the ... to ensure that each version of a preprint is assigned a unique DOI.

Comment (NateWr) [reply locale/en_US/manager.po:915]: This needs to go into the OPS repository.

**Review by NateWr — CHANGES_REQUESTED** [review]

**Review by bozana — COMMENTED** [review]
Also a few general comments/questions:

- I think the `doiVersioning` setting is not considered at the installation?
- In the function `unpublish()`  (publication repository), the `getDoisForSubmission()` is called: There only the current publication is considered (to mark stale). I think one could unpublish another, not the current publication as well, correct? (Maybe then it should have the unpublished publication as parameter and get those DOIs?)
- The documentation of the function `getDoisForSubmission()` in the code seems to be wrong
- Why do we not consider DOI status `unassigned` (instead of using `unregistered` for that too)? -- This way the user could maybe select an unregistered DOI and try to register it, although there is no DOI...

File: classes/publication/Repository.php
Comment (ewhanson) [reply classes/publication/Repository.php:299]: I'm not using it, but @NateWr suggested it be added now since we were already changing the function signature as it would likely be helpful to have the associated user for hooks/events/etc.

File: classes/components/listPanels/PKPDoiListPanel.php
Comment (ewhanson) [reply classes/components/listPanels/PKPDoiListPanel.php:180]: Looks like this is only in OPS, so I'll move it over to pkp-lib. I'll probably wait until right before merge to avoid translation conflicts. @asmecher, could I give you a heads up then for you to pull in the latest translations?

File: classes/publication/Repository.php
Comment (NateWr) [reply classes/publication/Repository.php:299]: @asmecher has pushed back on this idea in our discussion about events here: https://github.com/pkp/pkp-lib/discussions/8388#discussioncomment-4040677

Maybe until we settle on an approach, it's better to load the dependencies you need inside the event listener. It won't be as performant but it is easier to add arguments to a function later than it is to remove them.

Comment (bozana) [reply classes/publication/Repository.php:299]: This is also a place in the code where there is no event... So yes, I would remove it...

File: api/v1/dois/PKPDoiHandler.php
Comment (ewhanson) [reply api/v1/dois/PKPDoiHandler.php:289]: This one was a bit messy. It's not always included because this `edit()` function covers two (theoretical) use cases.

1. Being used as a "pure" API endpoint to only edit the DOI with no side effects. (This is the theoretical case, and I'm not sure if this should be allowed as it could lead to unintended consequences). This is the case without including `$pubObjectType` and `$pubObjectId`.
2. Being used to edit the DOIs in the context of their publication object (how it's used in the UI). The code below are the side effects that need to happen when editing a DOI in context. 

```PHP
// Copy DOI object data
$newDoi = clone $doi;
$newDoi->unsetData('id');
$newDoi->setAllData(array_merge($newDoi->getAllData(), ['doi' => $params['doi']]));
$newDoiId = Repo::doi()->add($newDoi);

// Update pubObject with new DOI and remove elsewhere if no longer in use
$pubObjectRepo->edit($pubObject, ['doiId' => $newDoiId]);
if (!Repo::doi()->isAssigned($doi->getId())) {
    Repo::doi()->delete($doi);
}
```

Given this, do you a request should always be required to include a `$pubObjectType` and `$pubObjectId` or is it okay to keep as is?

Comment (bozana) [reply api/v1/dois/PKPDoiHandler.php:289]: So the 1. case is maybe when editing a DOI manually via the UI/DOI management page?
And the 2. case is e.g. when publishing a new version?
If so maybe to add a comment in the function documentation or here in the code...
And, yes, we can then keep it as is...

**Comment by ewhanson** [comment]
Hi @bozana, thanks for reviewing this! To your specific questions:

> I think the `doiVersioning` setting is not considered at the installation?

- `doiVersioning` is part of the `context.json` schema file, so it should be setup correctly as part of the installation. Is there anywhere else it was missing?

> In the function `unpublish()` (publication repository), the `getDoisForSubmission()` is called: There only the current publication is considered (to mark stale). I think one could unpublish another, not the current publication as well, correct? (Maybe then it should have the unpublished publication as parameter and get those DOIs?)

- I've updated `getDoisForSubmission()` to account for all publications.

> The documentation of the function `getDoisForSubmission()` in the code seems to be wrong

- Thanks! Updated this as well.

> Why do we not consider DOI status `unassigned` (instead of using `unregistered` for that too)? -- This way the user could maybe select an unregistered DOI and try to register it, although there is no DOI...

- In terms of the possible statuses a DOI can have, there is no `unassigned` status because a DOI object won't exist until it has been assigned. In terms of the DOI management UI, I see how this could be confusing. We have a "Needs DOI" filter that partially addresses this use case, but maybe it would be helpful to have the badge that says "Unregistered" say "Unassigned" when there are no DOIs assigned. If that sounds like a good strategy to you, I can create a new issue for it.

**Comment by bozana** [comment]
Hi @ewhanson, thanks a lot!
If there is no reason not to do it I would prefer to have a badge that says "Needs DOI" (or "Unassigned")...

File: classes/doi/Repository.php
Comment (bozana) [inline classes/doi/Repository.php:315]: Do we need to implement this function in apps, e.g. for issues (ojs), files (omp)?

File: api/v1/dois/PKPDoiHandler.php
Comment (bozana) [inline api/v1/dois/PKPDoiHandler.php:318]: We could theoretically provide pubObjectType to the isAssigned function -- the same DOI can only be assigned to an object from the same type -- or?

File: classes/components/listPanels/PKPDoiListPanel.php
Comment (bozana) [inline classes/components/listPanels/PKPDoiListPanel.php:180]: We need to remove this key from here, right?

File: api/v1/dois/PKPDoiHandler.php
Comment (bozana) [inline api/v1/dois/PKPDoiHandler.php:292]: I think we need to consider issue type here -- it always fits into this default behavior because there is no versioning for issues -- correct? Else, further down the type not found will be returned when an issue DOI is edited.

File: api/v1/dois/PKPDoiHandler.php
Comment (bozana) [inline api/v1/dois/PKPDoiHandler.php:44]: I think we need to consider chapter and submission file type in OMP

File: classes/doi/Repository.php
Comment (ewhanson) [reply classes/doi/Repository.php:315]: Added!

File: classes/components/listPanels/PKPDoiListPanel.php
Comment (ewhanson) [reply classes/components/listPanels/PKPDoiListPanel.php:185]: No, this is the locale key that needed to be transferred from OPS to pkp-lib. This is the `"Version 1"`, `"Version 2"`, etc string used in the version modal.

File: api/v1/dois/PKPDoiHandler.php
Comment (ewhanson) [reply api/v1/dois/PKPDoiHandler.php:318]: That's a good idea. It would reduce the number of database queries for each check if it's only checking the one pub object type.

File: api/v1/dois/PKPDoiHandler.php
Comment (ewhanson) [reply api/v1/dois/PKPDoiHandler.php:44]: Yes, I've opted to use the preexisting constants in the DOI repository, e.g. `Repo::doi()::TYPE_PUBLICATION`.

File: api/v1/dois/PKPDoiHandler.php
Comment (ewhanson) [reply api/v1/dois/PKPDoiHandler.php:292]: Yes, I've rewritten it to account for issues in OJS so the issue type is now accounted for.

**Pull request merged**
